File: ml/src/semantic_dark_ml/manifest.py
from __future__ import annotations

import hashlib
import json
import re
from dataclasses import asdict, dataclass
from pathlib import Path, PurePosixPath
from typing import Iterable, Literal, Self

SCHEMA = "semantic-dark.corpus.v1"
CorpusLabel = Literal["photo", "icon", "diagram", "screenshot", "unknown"]
TargetSplit = Literal["train", "val", "test"]
_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._:-]*$")
_SHA256 = re.compile(r"^[0-9a-f]{64}$")


@dataclass(frozen=True, slots=True)
class CorpusRecord:
    id: str
    label: CorpusLabel
    source: str
    source_group: str
    target_split: TargetSplit
    path: str
    sha256: str
    raw_sha256: str
    original_width: int
    original_height: int
    license: str
    revision: str
    schema: str = SCHEMA

    def validate(self) -> None:
        if self.schema != SCHEMA:
            raise ValueError(f"Unsupported corpus schema: {self.schema}")
        if not _ID.fullmatch(self.id):
            raise ValueError(f"Invalid record id: {self.id!r}")
        if self.label not in {"photo", "icon", "diagram", "screenshot", "unknown"}:
            raise ValueError(f"Invalid label: {self.label}")
        if self.target_split not in {"train", "val", "test"}:
            raise ValueError(f"Invalid target split: {self.target_split}")
        relative = PurePosixPath(self.path)
        if relative.is_absolute() or ".." in relative.parts or not relative.parts:
            raise ValueError(f"Manifest path must be safe and relative: {self.path}")
        if not _SHA256.fullmatch(self.sha256) or not _SHA256.fullmatch(self.raw_sha256):
            raise ValueError(f"Invalid normalized or raw sha256 for {self.id}")
        if self.original_width <= 0 or self.original_height <= 0:
            raise ValueError(f"Invalid original dimensions for {self.id}")
        for name, value in (
            ("source", self.source),
            ("source_group", self.source_group),
            ("license", self.license),
            ("revision", self.revision),
        ):
            if not value.strip():
                raise ValueError(f"{name} is required for {self.id}")

    def to_dict(self) -> dict[str, object]:
        self.validate()
        return asdict(self)

    @classmethod
    def from_dict(cls, value: dict[str, object]) -> Self:
        record = cls(**value)  # type: ignore[arg-type]
        record.validate()
        return record


def write_manifest(records: Iterable[CorpusRecord], path: str | Path) -> int:
    manifest = Path(path)
    manifest.parent.mkdir(parents=True, exist_ok=True)
    count = 0
    with manifest.open("w", encoding="utf-8", newline="\n") as handle:
        for record in records:
            handle.write(json.dumps(record.to_dict(), sort_keys=True, separators=(",", ":")) + "\n")
            count += 1
    return count


def load_manifest(path: str | Path, *, verify_files: bool = False) -> list[CorpusRecord]:
    manifest = Path(path)
    records: list[CorpusRecord] = []
    seen_ids: set[str] = set()
    with manifest.open(encoding="utf-8") as handle:
        for line_number, line in enumerate(handle, 1):
            if not line.strip():
                continue
            try:
                value = json.loads(line)
                if not isinstance(value, dict):
                    raise TypeError("record is not an object")
                record = CorpusRecord.from_dict(value)
            except (TypeError, ValueError, json.JSONDecodeError) as error:
                raise ValueError(f"Invalid manifest line {line_number}: {error}") from error
            if record.id in seen_ids:
                raise ValueError(f"Duplicate manifest id: {record.id}")
            seen_ids.add(record.id)
            if verify_files:
                file_path = (manifest.parent / record.path).resolve()
                if not file_path.is_relative_to(manifest.parent.resolve()):
                    raise ValueError(f"Record path escapes manifest root: {record.path}")
                if sha256_file(file_path) != record.sha256:
                    raise ValueError(f"Checksum mismatch: {record.path}")
            records.append(record)
    return records


def sha256_file(path: str | Path) -> str:
    digest = hashlib.sha256()
    with Path(path).open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()

File: src/testing/vision-benchmark/manifest.ts
import {createHash} from 'node:crypto';
import {readFile} from 'node:fs/promises';
import path from 'node:path';
import {PNG} from 'pngjs';
import type {RGBAImage} from '../../vision';
import {
  KNOWN_LABELS,
  type CorpusLabel,
  type CorpusManifestRecord,
  type LocatedCorpusRecord,
  type TargetSplit,
} from './types';

const LABELS = new Set<CorpusLabel>([...KNOWN_LABELS, 'unknown']);
const SPLITS = new Set<TargetSplit>(['train', 'val', 'test']);
const SHA256 = /^[0-9a-f]{64}$/;

export async function loadCorpusManifests(
  manifestPaths: readonly string[],
  split?: TargetSplit,
): Promise<LocatedCorpusRecord[]> {
  const records: LocatedCorpusRecord[] = [];
  const seenIds = new Set<string>();
  for (const manifestValue of manifestPaths) {
    const manifest = path.resolve(manifestValue);
    const root = path.dirname(manifest);
    const text = await readFile(manifest, 'utf8');
    for (const [index, line] of text.split(/\r?\n/).entries()) {
      if (!line.trim()) continue;
      const record = parseRecord(JSON.parse(line) as unknown, manifest, index + 1);
      if (seenIds.has(record.id)) throw new Error(`Duplicate corpus id: ${record.id}`);
      seenIds.add(record.id);
      if (split && record.target_split !== split) continue;
      const absolutePath = path.resolve(root, record.path);
      if (absolutePath !== root && !absolutePath.startsWith(`${root}${path.sep}`)) {
        throw new Error(`Corpus path escapes manifest root: ${record.path}`);
      }
      records.push({record, absolutePath});
    }
  }
  return records;
}

export async function readVerifiedPng(item: LocatedCorpusRecord): Promise<RGBAImage> {
  const bytes = await readFile(item.absolutePath);
  const digest = createHash('sha256').update(bytes).digest('hex');
  if (digest !== item.record.sha256) {
    throw new Error(`Checksum mismatch for ${item.record.id}`);
  }
  const decoded = PNG.sync.read(bytes);
  return {
    width: decoded.width,
    height: decoded.height,
    data: new Uint8ClampedArray(decoded.data),
  };
}

/** Reject exact source or normalized-content reuse across named splits. */
export function assertCorpusDisjoint(records: readonly LocatedCorpusRecord[]): void {
  const groups = new Map<string, Set<TargetSplit>>();
  for (const {record} of records) {
    for (const key of [
      `source-group:${record.source_group}`,
      `normalized-sha256:${record.sha256}`,
      `raw-sha256:${record.raw_sha256}`,
    ]) {
      const splits = groups.get(key) ?? new Set<TargetSplit>();
      splits.add(record.target_split);
      groups.set(key, splits);
    }
  }
  const leaks = [...groups.entries()]
    .filter(([, splits]) => splits.size > 1)
    .map(([key, splits]) => `${key} -> ${[...splits].sort().join(',')}`);
  if (leaks.length > 0) {
    throw new Error(`Corpus leakage detected: ${leaks.slice(0, 5).join('; ')}`);
  }
}

function parseRecord(value: unknown, manifest: string, line: number): CorpusManifestRecord {
  if (!value || typeof value !== 'object') throw invalid(manifest, line, 'row is not an object');
  const row = value as Record<string, unknown>;
  if (row.schema !== 'semantic-dark.corpus.v1') throw invalid(manifest, line, 'bad schema');
  for (const key of [
    'id',
    'source',
    'source_group',
    'path',
    'sha256',
    'raw_sha256',
    'license',
    'revision',
  ] as const) {
    if (typeof row[key] !== 'string' || !row[key].trim()) {
      throw invalid(manifest, line, `${key} is required`);
    }
  }
  if (!LABELS.has(row.label as CorpusLabel)) throw invalid(manifest, line, 'bad label');
  if (!SPLITS.has(row.target_split as TargetSplit)) throw invalid(manifest, line, 'bad split');
  if (!SHA256.test(row.sha256 as string) || !SHA256.test(row.raw_sha256 as string)) {
    throw invalid(manifest, line, 'bad normalized or raw sha256');
  }
  if (!Number.isInteger(row.original_width) || (row.original_width as number) <= 0 ||
      !Number.isInteger(row.original_height) || (row.original_height as number) <= 0) {
    throw invalid(manifest, line, 'bad original dimensions');
  }
  const relative = row.path as string;
  if (path.isAbsolute(relative) || relative.split(/[\\/]/).includes('..')) {
    throw invalid(manifest, line, 'path must be safe and relative');
  }
  return row as unknown as CorpusManifestRecord;
}

function invalid(manifest: string, line: number, reason: string): Error {
  return new Error(`Invalid manifest ${manifest}:${line}: ${reason}`);
}
